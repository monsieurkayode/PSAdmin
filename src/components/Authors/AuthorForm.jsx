import React from 'react';
import {
  shape, string, func, bool
} from 'prop-types';
import TextInput from '../common/TextInput/Index';

const AuthorForm = ({
  firstName,
  lastName,
  errors,
  onChange,
  onSubmit,
  onFocus,
  saving,
  loading
}) => (
  <form onSubmit={onSubmit}>
    <TextInput
      name="firstName"
      label="First Name"
      placeholder="Kent"
      onChange={onChange}
      onFocus={onFocus}
      loading={loading}
      value={firstName}
      error={errors.firstName}
    />
    <TextInput
      name="lastName"
      label="Last Name"
      placeholder="Dodds"
      onChange={onChange}
      onFocus={onFocus}
      loading={loading}
      value={lastName}
      error={errors.lastName}
    />
    <input
      type="submit"
      value={saving ? 'Saving...' : 'Save'}
      disabled={saving}
      className="btn btn-primary"
    />
  </form>
);

AuthorForm.defaultProps = {
  firstName: '',
  lastName: '',
  errors: {},
};

AuthorForm.propTypes = {
  firstName: string,
  lastName: string,
  errors: shape({}),
  saving: bool.isRequired,
  loading: bool.isRequired,
  onChange: func.isRequired,
  onSubmit: func.isRequired,
  onFocus: func.isRequired
};

export default AuthorForm;
