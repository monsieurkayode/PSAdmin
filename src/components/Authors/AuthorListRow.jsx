import React from 'react';
import { Link } from 'react-router-dom';
import { string, func } from 'prop-types';

const AuthorListRow = ({
  id,
  firstName,
  lastName,
  deleteAuthor
}) => (
  <tr>
    <td>&nbsp;</td>
    <td><Link to={`/author/${id}`}>{id}</Link></td>
    <td>{firstName}&nbsp;{lastName}</td>
    <td>
      <button
        type="button"
        className="btn btn-sm btn-danger"
        onClick={() => deleteAuthor(id)}
      >
        Delete
      </button>
    </td>
  </tr>
);

AuthorListRow.propTypes = {
  id: string.isRequired,
  firstName: string.isRequired,
  lastName: string.isRequired,
  deleteAuthor: func.isRequired
};

export default AuthorListRow;
