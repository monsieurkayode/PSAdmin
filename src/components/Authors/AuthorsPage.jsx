import React from 'react';
import { Link } from 'react-router-dom';

const AuthorsPage = () => (
  <div className="container jumbotron">
    <div className="container-fluid">
      <h1>Authors</h1>
      <p>Authors management page.</p>
      <Link to="/authors">
        <button type="button" className="btn btn-lg btn-primary">
          Add Author
        </button>
      </Link>
    </div>
  </div>
);

export default AuthorsPage;
