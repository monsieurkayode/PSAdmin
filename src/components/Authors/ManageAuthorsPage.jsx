import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  shape,
  bool,
  string,
  func
} from 'prop-types';
import toastr from 'toastr';
import { Prompt } from 'react-router-dom';
import { saveAuthor, setSaveErrors } from '../../actions/authorActions';
import AuthorForm from './AuthorForm';
import NotFoundPage from '../NotFound/NotFoundPage';
import { validateInput, checkChanges } from '../../helpers/validator';
import getAuthor from '../../helpers/getData';


export class ManageAuthorsPage extends Component {
  static defaultProps = {
    author: {},
    match: {}
  }

  static propTypes = {
    author: shape({}),
    isSaving: bool.isRequired,
    loadingAuthor: bool.isRequired,
    history: shape({
      push: func.isRequired
    }).isRequired,
    match: shape({
      params: {
        id: string
      }
    })
  }

  static initialState = () => ({
    author: {
      id: '',
      firstName: '',
      lastName: ''
    },
    errors: {},
    newChanges: false
  })

  static getDerivedStateFromProps(
    { author, isSaving }, { errors, newChanges }
  ) {
    if (!isSaving && !newChanges) {
      return {
        ...ManageAuthorsPage.initialState(),
        author,
        errors,
        newChanges
      };
    }

    return null;
  }

  state = ManageAuthorsPage.initialState();

  redirect = this.props.history.push;

  componentWillUnmount() {
    this.props.setSaveErrors({});
  }

  handleInputChange = (event) => {
    const { author } = this.state;
    author[event.target.name] = event.target.value;
    this.setState({ author, newChanges: checkChanges(author) });
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { author } = this.state;
    const { isValid, errors } = validateInput(author, false);

    if (!isValid) {
      toastr.clear();
      toastr.error('Invalid form data', 'Submission Error!');
      return this.setState({ errors });
    }

    this.setState({ newChanges: false });

    this.props.saveAuthor(author)
      .then(() => this.redirect('/authors'))
      .catch(err => this.setState({ errors: err }));
  }

  clearErrors = (event) => {
    toastr.clear();
    const { errors } = this.state;
    errors[event.target.name] = '';
    if (Object.keys(this.state.errors).length > 0) {
      this.setState({ errors });
    }
  }

  render() {
    const { author, errors, newChanges } = this.state;
    const { isSaving, loadingAuthor, match } = this.props;

    if (!loadingAuthor && match.params.id && !this.props.author.id) {
      return <NotFoundPage content="Author" />;
    }

    return (
      <div className="container" style={{ maxWidth: '55%', minWidth: '400px' }}>
        <Prompt
          when={!isSaving && newChanges}
          message="You have unsaved changes, leave page?"
        />
        <h1>Manage Author</h1>
        <AuthorForm
          {...author}
          errors={errors}
          onChange={this.handleInputChange}
          onSubmit={this.handleSubmit}
          onFocus={this.clearErrors}
          saving={isSaving}
          loading={loadingAuthor}
        />
      </div>
    );
  }
}

/**
 *
 * @description makes the ManageAuthorsPage a container component
 * by making it aware of the application state in the store
 * via the connect function from react-redux library
 *
 * @function mapStateToProps
 *
 * @param {object} { allAuthors }
 */
const mapStateToProps = ({
  allAuthors: {
    authors, isLoading, isSaving, errors
  }
}, { match: { params } }) => {
  let { author } = ManageAuthorsPage.initialState();

  if (getAuthor(authors, params.id)) {
    author = Object.assign({}, getAuthor(authors, params.id));
  }

  let loadingAuthor = isLoading;

  if (!params.id) { loadingAuthor = false; }

  return {
    isSaving,
    loadingAuthor,
    errors,
    authors,
    author
  };
};

export default connect(
  mapStateToProps,
  { saveAuthor, setSaveErrors }
)(ManageAuthorsPage);
