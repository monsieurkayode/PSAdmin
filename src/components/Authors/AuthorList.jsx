import React from 'react';
import { arrayOf, shape, func } from 'prop-types';
import AuthorListRow from './AuthorListRow';

const AuthorList = ({ authors, deleteAuthor }) => (
  <table className="table">
    <thead>
      <tr>
        <th>&nbsp;</th>
        <th>ID</th>
        <th>Author Name</th>
        <th>&nbsp;</th>
      </tr>
    </thead>
    <tbody>
      {authors.map(author => (
        <AuthorListRow
          key={author.id}
          {...author}
          deleteAuthor={deleteAuthor}
        />
      ))}
    </tbody>
  </table>
);

AuthorList.propTypes = {
  authors: arrayOf(shape({})).isRequired,
  deleteAuthor: func.isRequired,
};

export default AuthorList;
