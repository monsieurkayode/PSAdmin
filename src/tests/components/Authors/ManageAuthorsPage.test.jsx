import React from 'react';
import { shallow } from 'enzyme';
import { ManageAuthorsPage } from '../../../components/Authors/ManageAuthorsPage';
import mockAuthor from '../../__mocks__/mockAuthor';

const setup = () => {
  const props = {
    author: mockAuthor,
    isSaving: false,
    loadingAuthor: false,
    history: {
      push: jest.fn()
    },
    saveAuthor: () => new Promise(resolve => resolve(mockAuthor)),
    setSaveErrors: jest.fn(),
    match: {
      params: {
        id: mockAuthor.id
      }
    }
  };

  const state = {
    errors: {},
    newChanges: false
  };

  const wrapper = shallow(<ManageAuthorsPage {...props} />);

  return {
    wrapper,
    props,
    state
  };
};

const event = {
  target: {
    name: 'firstName',
    value: ''
  },
  preventDefault: jest.fn()
};

describe('ManageAuthorsPage Component', () => {
  const { wrapper, state, props: { isSaving, author } } = setup();

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a h1 tag element', () => {
    expect(wrapper.find('h1').length).toBe(1);
    expect(wrapper.find('h1').text()).toEqual('Manage Author');
  });

  it('renders a Prompt child component', () => {
    expect(wrapper.find('Prompt').length).toBe(1);
  });

  it('renders a AuthorForm child component', () => {
    expect(wrapper.find('AuthorForm').length).toBe(1);
  });

  it('invokes the static method getDerivedStateFromProps', () => {
    const { errors, newChanges } = state;
    const result = ManageAuthorsPage
      .getDerivedStateFromProps({ author, isSaving }, { errors, newChanges });
    expect(result.author).toEqual(author);
  });

  it('invokes handleSubmit method', () => {
    const handleSubmitSpy = jest.spyOn(wrapper.instance(), 'handleSubmit');
    wrapper.instance().handleSubmit(event);
    expect(handleSubmitSpy).toHaveBeenCalled();
  });

  it('invokes clearErrors method', () => {
    const clearErrorsSpy = jest.spyOn(wrapper.instance(), 'clearErrors');
    wrapper.instance().clearErrors(event);
    expect(clearErrorsSpy).toHaveBeenCalled();
  });

  it('invokes handleInputChange method', () => {
    const handleInputChangeSpy = jest
      .spyOn(wrapper.instance(), 'handleInputChange');
    wrapper.instance().handleInputChange(event);
    expect(handleInputChangeSpy).toHaveBeenCalled();
  });

  it('invokes componentWillMount method', () => {
    jest.spyOn(wrapper.instance(), 'componentWillUnmount');
    wrapper.instance().componentWillUnmount();
    expect(wrapper.instance().componentWillUnmount).toHaveBeenCalled();
  });
});
