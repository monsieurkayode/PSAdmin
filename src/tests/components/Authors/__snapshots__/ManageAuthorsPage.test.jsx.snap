// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageAuthorsPage Component renders correctly 1`] = `
<div
  className="container"
  style={
    Object {
      "maxWidth": "55%",
      "minWidth": "400px",
    }
  }
>
  <Prompt
    message="You have unsaved changes, leave page?"
    when={false}
  />
  <h1>
    Manage Author
  </h1>
  <AuthorForm
    errors={Object {}}
    firstName="Cory"
    id="cory-house"
    lastName="House"
    loading={false}
    onChange={[Function]}
    onFocus={[Function]}
    onSubmit={[Function]}
    saving={false}
  />
</div>
`;
